File: backend/app/api/v1/fx_rates.py
"""FX Rates API – manual entry + live fetch from frankfurter.app (free, no key needed)."""
import httpx
from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.core.security import get_current_user
from app.models.fx_rate import FxRate
from pydantic import BaseModel
from datetime import date, timedelta
from decimal import Decimal

router = APIRouter()


class FxRateCreate(BaseModel):
    rate_date : date
    usd_jpy   : Decimal
    rate_type : str = "MUFG_TTM"


@router.get("/")
def list_rates(db: Session = Depends(get_db)):
    return db.query(FxRate).order_by(FxRate.rate_date.desc()).limit(60).all()


@router.get("/latest")
def latest_rate(db: Session = Depends(get_db)):
    rate = db.query(FxRate).order_by(FxRate.rate_date.desc()).first()
    if rate:
        return {"date": str(rate.rate_date), "usd_jpy": float(rate.usd_jpy),
                "source": rate.source or "manual"}
    return {"date": None, "usd_jpy": None, "source": None}


@router.get("/live")
def live_rate():
    """Fetch live USD/JPY from frankfurter.app (ECB data – free, no API key)."""
    try:
        resp = httpx.get(
            "https://api.frankfurter.dev/v1/latest?from=USD&to=JPY",
            timeout=8.0,
            follow_redirects=True,
        )
        resp.raise_for_status()
        data = resp.json()
        rate = data["rates"]["JPY"]
        return {
            "date"   : data["date"],
            "usd_jpy": rate,
            "source" : "frankfurter.app (ECB)",
        }
    except Exception as e:
        raise HTTPException(status_code=503, detail=f"Could not fetch live rate: {e}")


@router.get("/history")
def rate_history(days: int = 30, db: Session = Depends(get_db)):
    since = date.today() - timedelta(days=days)
    rates = (
        db.query(FxRate)
        .filter(FxRate.rate_date >= since)
        .order_by(FxRate.rate_date.asc())
        .all()
    )
    return [{"date": str(r.rate_date), "usd_jpy": float(r.usd_jpy), "source": r.source or "manual"} for r in rates]


@router.post("/")
def create_rate(rate: FxRateCreate,
                db: Session = Depends(get_db),
                user = Depends(get_current_user)):
    existing = db.query(FxRate).filter(FxRate.rate_date == rate.rate_date).first()
    if existing:
        existing.usd_jpy   = rate.usd_jpy
        existing.rate_type = rate.rate_type
        db.commit()
        return existing
    db_rate = FxRate(**rate.model_dump())
    db.add(db_rate)
    db.commit()
    db.refresh(db_rate)
    return db_rate

File: backend/src/routes/fxRates.ts
/** FX Rates — /api/v1/fx-rates */

import { Hono } from 'hono'
import { prisma } from '../lib/prisma'
import { auth, type AuthVars } from '../middleware/auth'

const app = new Hono<AuthVars>()
app.use('*', auth)

function rateDict(r: any) {
  return {
    id:       r.id,
    date:     r.rateDate?.toISOString().slice(0, 10),
    usd_jpy:  parseFloat(r.usdJpy.toString()),
    source:   r.source,
  }
}

// GET /
app.get('/', async (c) => {
  const rates = await prisma.fxRate.findMany({ orderBy: { rateDate: 'desc' } })
  return c.json(rates.map(rateDict))
})

// GET /latest
app.get('/latest', async (c) => {
  const rate = await prisma.fxRate.findFirst({ orderBy: { rateDate: 'desc' } })
  if (!rate) return c.json({ usd_jpy: null, date: null })
  return c.json({ usd_jpy: parseFloat(rate.usdJpy.toString()), date: rate.rateDate.toISOString().slice(0, 10) })
})

// GET /live  (fetches from frankfurter.app)
app.get('/live', async (c) => {
  try {
    const res  = await fetch('https://api.frankfurter.app/latest?from=USD&to=JPY')
    const data = await res.json() as any
    const rate = data?.rates?.JPY
    if (!rate) return c.json({ detail: 'Live rate unavailable' }, 503)
    return c.json({ usd_jpy: rate, date: data.date })
  } catch {
    return c.json({ detail: 'Live rate fetch failed' }, 503)
  }
})

// GET /history
app.get('/history', async (c) => {
  const days   = parseInt(c.req.query('days') ?? '90')
  const cutoff = new Date()
  cutoff.setDate(cutoff.getDate() - days)

  const rates = await prisma.fxRate.findMany({
    where:   { rateDate: { gte: cutoff } },
    orderBy: { rateDate: 'asc' },
  })
  return c.json(rates.map(rateDict))
})

// POST /
app.post('/', async (c) => {
  const { rate_date, usd_jpy, source = 'manual' } = await c.req.json().catch(() => ({}))
  if (!rate_date || !usd_jpy) return c.json({ detail: 'rate_date and usd_jpy required' }, 400)

  // Upsert — one rate per date
  const existing = await prisma.fxRate.findFirst({ where: { rateDate: new Date(rate_date) } })
  if (existing) {
    const updated = await prisma.fxRate.update({
      where: { id: existing.id },
      data:  { usdJpy: usd_jpy, source },
    })
    return c.json(rateDict(updated))
  }

  const rate = await prisma.fxRate.create({
    data: { rateDate: new Date(rate_date), usdJpy: usd_jpy, source },
  })
  return c.json(rateDict(rate))
})

export default app
